nixos/modules/installer/netboot/kexec-boot.sh

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env bash
+set -e
+
+kexecTree=$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)
+initrd=$kexecTree/initrd
+kernelParams=$(<"$kexecTree/kernel-params")
+
+usage() {
+    printf >&2 'usage: %s [%s] [%s]\n' "$0" \
+        '--ssh-authorized-keys <file>' \
+        '--append <cmdline>'
+    exit 2
+}
+
+while (( $# )); do
+    arg=$1; shift 1
+    case "$arg" in
+        --ssh-authorized-keys)
+            (( $# )) || usage
+            authorizedKeysFile=$1; shift 1
+            ;;
+        --append)
+            (( $# )) || usage
+            kernelParams="$kernelParams $1"; shift 1
+            ;;
+        *)
+            usage
+            ;;
+    esac
+done
+
+if ! kexec --version &>/dev/null; then
+    printf >&2 'kexec not found: please install kexec-tools\n'
+    exit 1
+fi
+
+if (( EUID != 0 )); then
+    printf >&2 'kexec-boot must be run as root\n'
+    exit 1
+fi
+
+cleanup() {
+    if [[ -v workDir ]]; then
+        rm -rf "$workDir"
+    fi
+}
+
+trap cleanup EXIT
+
+if [[ -v authorizedKeysFile ]]; then
+    workDir=$(mktemp -d --tmpdir nixos-kexec-initrd.XXXXXXXXXX)
+    keysDir=$workDir/overlay/etc/ssh/authorized_keys.d
+    mkdir -p -- "$keysDir"
+    install --mode=0644 -- "$authorizedKeysFile" "$keysDir/nixos"
+    {
+        (
+            cd -- "$workDir"
+            find overlay -print0 | cpio -o -H newc --null --quiet
+        )
+        cat -- "$initrd"
+    } >"$workDir/initrd"
+    initrd="$workDir/initrd"
+fi
+
+kexec --load \
+    --kexec-syscall-auto \
+    --initrd="$initrd" \
+    --append="$kernelParams" \
+    -- "$kexecTree/kernel"
+
+cleanup
+
+if [[ -d /run/systemd/system ]] && command -v systemctl >/dev/null; then
+    systemctl kexec
```

Comment:
@mlyxshi in those cases where it failed to load, would it brick the system? If not we could just try a normal kexec in case `systemctl kexec` fails.

Reply:
My environment:
Azure b1s debian Gen2 x86_64

It is hard for me to debug why `systemctl kexec` do not work. I switch to normal `kexec -e`

Reply:
But the question is if this leaves the machine in an unusable state or if this just fails and one can run different commands?

Reply:
`systemctl kexec` should just be performing a normal shutdown procedure and then invoking the same system calls that `kexec(8)` does. All I can think of is that it perhaps uninitializes some hardware that breaks the kernel boot? But then the NixOS kernel probably wouldn't work without kexec, either.

Reply:
Another problem is the following:

```
[root@nixos:~]# sudo systemctl kexec
No kexec kernel loaded and autodetection failed.
Automatic loading works only on systems booted with EFI.
```

On providers like Hetzner and some cloud providers there is no EFI support.

Reply:
I'm pretty sure this only applies to automatically determining the kernel to load with the UEFI [Boot Loader Interface](https://systemd.io/BOOT_LOADER_INTERFACE/) (hence "Automatic loading works only ..."). We load the kernel manually ourselves before calling `systemctl kexec` so it shouldn't be a problem.